tests/statespace_test.h

```diff
@@ -387,14 +351,92 @@ R"(20
 30 y_1_2 19
 )";
 
+template <typename Simulator>
+void TestNormAndInnerProduct() {
+  unsigned depth = 8;
+
+  std::stringstream ss(circuit_string);
+  Circuit<GateQSim<float>> circuit;
+  EXPECT_EQ(CircuitReader<IO>::FromStream(depth, provider, ss, circuit), true);
+  circuit.gates.emplace_back(GateT<float>::Create(depth + 1, 0));
+
+  using StateSpace = typename Simulator::StateSpace;
+  using State = typename StateSpace::State;
+  using Runner = QSimRunner<IO, BasicGateFuser<GateQSim<float>>, Simulator>;
+
+  StateSpace state_space(circuit.num_qubits, 1);
+  State state0 = state_space.CreateState();
+
+  EXPECT_EQ(state_space.IsNull(state0), false);
+
+  auto measure = [&state0](unsigned k,
```

Comment:
As I see it, this is taking the states at time `depth` and time `depth + 1` and checking their inner product - is that correct?

Reply:
Yes, that is correct. There is only one T gate at time `depth + 1`. So the inner product can be estimated analytically: {(1 + cos(pi/4))/2, sin(pi/4)/2}.